[Feat] document EBR use case (#53)
libremfg/libremfg.github.io: content/use-cases/ebr.md

---
title: >-
  Electronic batch records
description: An example of how to use Rhize to create Electronic Batch Records for pharmaceutical manufacturing
categories: ["howto", "use-cases"]
weight: 001
menu:
  main:
    parent: use-cases
    identifier:
---

In pharmaceutical manufacturing, an _Electronic Batch Record (eBR)_ is a detailed report about a specific batch.
It includes information about items such as the process segments, material, personnel, and equipment.

As Rhize can ingest data from all levels of the manufacturing process, and do so in real-time, it is an ideal single source of truth to create eBRs.
The process can be automated and highly efficient.
Our experience in production systems shows that Rhize can typically generate an eBR in less than 500 milliseconds. 

The procedure has the following steps:

1. Identify the sources of data.
1. Map the fields for these data sources to Rhize's ISA-95 schema.

> 1. Write a {{< abbr "BPMN" >}} process that listens for a data source, transforms the incoming data to the schema, and then sends a {{< abbr "mutation" >}} to update the Rhize database.

Reviewer: It is very rare that this will be only 1 process. Suggest we change to:

`1. Write {{< abbr "BPMN" >}} processes that listen to data sources, transforms the incoming data to the schema, and then sends a {{< abbr "mutatino" >}} to update the Rhize database.

> 1. Write a {{< abbr "BPMN" >}} process that listens for a data source, transforms the incoming data to the schema, and then sends a {{< abbr "mutation" >}} to update the Rhize database.

Author: ```suggestion
1. Write {{< abbr "BPMN" >}} processes that listen to data sources, transform the incoming data to the schema, and then sends a {{< abbr "mutation" >}} to update the Rhize database.
```

{{< figure
alt="Screenshot of a BPMN workflow that adds material only if it exists"
src="/images/bpmn/screenshot-rhize-bpmn-add-material-definition.png"
>}}
</a>

This step can involve multiple BPMN processes subscribing to different topics.
As long as the incoming event data has a common association, for example through the `id` of the batch data and associated `JobResponse`, you can return all eBR fields in one GraphQL query&mdash;no recursive SQL joins needed.

{{< figure
alt="Multiple BPMN processes can be united in one batch"
src="/images/ebr/diagram-rhize-inputs-for-ebr.png"
width="75%"
>}}

### Query the DB with the eBR fields

After the batch finishes, use a [GraphQL query]({{< relref "how-to/gql/query-mutate-subscribe" >}}) to receive all relevant batch data.
You only need to write one request, and it returns exactly the data you specify.

{{< figure
alt="Diagram showing how a query makes an ebr"
src="/images/ebr/diagram-rhize-make-ebr-query.png"
width="55%"
>}}

Our experience has shown us that the query is about 300 lines. Here's a
small, generic snippet of how it looks:

{{< expandable title="Snippet of a makeEbr query" >}}
```graphql

> query makeEbr ($filter: JobOrderFilter) {

Reviewer: This snippet is of the planned execution, not the actual. I'd suggest changing it to:

```graphql
query makeEbr ($filter: JobResponseFilter) {
  queryJobResponse(filter: $filter) {
    EXAMPLE_id: id
    description
    matActualProduced: materialActual(filter:{materialUse: { eq:Produced }}){
      id
      material: materialDefinitionVersion{
        id
      }
      quantity
      quantityUoM {
        id
      }
    }
  ## More EBR fields
  }
}
```

>   queryJobOrder(filter: $filter) {
>     EXAMPLE_id: id
>     label
>     matReqProduced: materialRequirements(filter: {materialUse:{eq: Produced}}){
>       id
>       material {
>     }
>     materialDefinitionVersion {
>       id
>       version
>       versionStatus
>     materialClasses {
>       id
>       label
>     }
>   ## More EBR fields
>   }

Author: ```suggestion
query makeEbr ($filter: JobResponseFilter) {
  queryJobResponse(filter: $filter) {
    EXAMPLE_id: id
    description
    matActualProduced: materialActual(filter:{materialUse: { eq:Produced }}){
      id
      material: materialDefinitionVersion{
        id
      }
      quantity
      quantityUoM {
        id
      }
    }
  ## More EBR fields
  }
}
```

---
title: >-
  Electronic batch records

> description: An example of how to use Rhize to create an Electronic Batch Records for pharmaceuticals

Author: ```suggestion
description: An example of how to use Rhize to create Electronic Batch Records for pharmaceuticals
```

> draft: false

Author: ```suggestion
draft: false
```

For example, your integrators may help appropriately model the data, and the frontend team may render the outputted JSON into a final document.

## Steps to automate eBR creation

The following steps describe the broad procedure to automatically update the database for incoming batch data, then automatically create an eBR after the batch run.

### Identify the sources of data

The first step is to identify the sources of data for your eBR.

{{< figure
alt="Diagram showing some examples of eBR data sources"
src="/images/ebr/diagram-rhize-example-ebr-sources.png"
width="75%"
>}}

Common data sources for an eBR might include:

- **{{< abbr "ERP" >}} documents:** high-level operations documents
- **{{< abbr "MES" >}} data:** granular process data, tracking objects such as the weight of individual material and the responses for different jobs.
- **{{< abbr "LIMS" >}} documents:** information about the laboratory environment
- **Real-time event data:** for example, data sent from OPC UA or MQTT servers

### Model the fields as ISA-95 schema

After you've identified the source data, the next step is to map this data into ISA-95 models.

> 

Author: ```suggestion
```

{{< figure
alt="Multiple BPMN processes can be united in one batch"
src="/images/ebr/diagram-rhize-inputs-for-ebr.png"
width="75%"
>}}

### Query the DB with the eBR fields

After the batch finishes, use a [GraphQL query]({{< relref "how-to/gql/query-mutate-subscribe" >}}) to receive all relevant batch data.
You only need to write one request, and it will return exactly the data you need.

{{< figure
alt="Diagram showing how a query makes an ebr"
src="/images/ebr/diagram-rhize-make-ebr-query.png"
width="55%"
>}}

Our experience has shown us that the query is about 300 lines. Here's a
small, generic snippet of how it looks:

{{< expandable title="Snippet of a makeEbr query" >}}
```graphql
query makeEbr ($filter: JobOrderFilter) {
  queryJobOrder(filter: $filter) {
    EXAMPLE_id: id
    label
    matReqProduced: materialRequirements(filter: {materialUse:{eq: Produced}}){
      id
      material {
    }
    materialDefinitionVersion {
      id
      version
      versionStatus
    materialClasses {
      id
      label
    }
  ## More EBR fields

>   }
> ```

Author: ```suggestion
  }
```

```
